File: python/testdata/fit_model.py
from scipy.linalg import block_diag
from scipy.linalg import solve
import numpy as np
# fit_model
def fit_model(Rxx, Ryy, ryy, gamma, base):
    # apply basis functions, if available 
    if base[0] is not None:
        B = block_diag(*base)
        Rxx = B.T @ Rxx @ B
        Rxy = B.T @ Rxy
        
    # Regularizer
    Gamma = gamma * np.diag(np.diag(Rxx)) # Tikhonov, scaled for all variables to be regularized equally, regardless of magnitude

    # Least squares estimate with regularization
    h = solve(Rxx + Gamma, Rxy)

    # mean error square
    Rxyest = Rxx @ h
    s2 = (np.sum(h * Rxyest) - 2 * np.sum(h * Rxy) + ryy).T

    # Bias term for ridge regression bias -- see Babadi derivation 
    Bias = np.sum((Rxy - Rxyest) * solve(Rxx, Rxy - Rxyest)).T / s2 / 2 if gamma > 0 else 0

    return h, s2, Bias

File: R/fit_model.R
library(MASS) # For generalized inverse functions
library(Matrix) # For dealing with sparse matrices

fit_model <- function(Rxx, Rxy, ryy, gamma, base) {
  # Apply basis functions if available
  if (!is.null(base[[1]])) {
    B <- do.call(bdiag, base)  # Construct a block diagonal matrix from the list of basis matrices
    Rxx <- t(B) %*% Rxx %*% B
    Rxy <- t(B) %*% Rxy
  }
  
  # Regularization
  Gamma <- gamma * diag(diag(Rxx)) # Tikhonov regularization scaled by diagonal elements of Rxx
  
  # Least squares estimate with regularization
  h <- solve(Rxx + Gamma, Rxy)
  
  # Mean error square
  Rxyest <- Rxx %*% h
  s2 <- rowSums(h * Rxyest) - 2 * rowSums(h * Rxy) + ryy
  
  # Bias term for ridge regression bias
  Bias <- if (gamma > 0) {
    residual <- Rxy - Rxyest
    rowSums((residual * solve(Rxx, residual)) / s2) / 2
  } else {
    rep(0, length(ryy))
  }
  
  return(list(h = h, s2 = s2, Bias = Bias))
}